libs/auth-api-lib/src/lib/delegations/delegations-incoming-custom.service.ts

```diff
@@ -0,0 +1,308 @@
+import { LOGGER_PROVIDER } from '@island.is/logging'
+import { isDefined } from '@island.is/shared/utils'
+import { partitionWithIndex } from './utils/partitionWithIndex'
+import { AuditService } from '@island.is/nest/audit'
+import { DelegationsOutgoingService } from './delegations-outgoing.service'
+import { Features } from '@island.is/feature-flags'
+import { FeatureFlagService } from '@island.is/nest/feature-flags'
+
+export const UNKNOWN_NAME = 'Óþekkt nafn'
+
+/**
+ * Service class for incoming delegations.
+ * This class supports domain based delegations.
+ */
+@Injectable()
+export class DelegationsIncomingCustomService {
+  constructor(
+    @InjectModel(Delegation)
+    private delegationModel: typeof Delegation,
+    @InjectModel(ClientAllowedScope)
+    private clientAllowedScopeModel: typeof ClientAllowedScope,
+    @Inject(DelegationConfig.KEY)
+    private delegationConfig: ConfigType<typeof DelegationConfig>,
+    private nationalRegistryClient: NationalRegistryClientService,
+    private featureFlagService: FeatureFlagService,
+    @Inject(LOGGER_PROVIDER)
+    private logger: Logger,
+    private auditService: AuditService,
+    private delegationsOutgoingService: DelegationsOutgoingService,
+  ) {}
+
+  async findAllValidIncoming(
+    user: User,
+    domainName?: string,
+  ): Promise<DelegationDTO[]> {
+    const feature = await this.featureFlagService.getValue(
+      Features.customDelegations,
+      false,
+      user,
+    )
+    if (!feature) {
+      return []
+    }
+
+    const { delegations, fromNameInfo } = await this.findAllIncoming(
+      user,
+      DelegationValidity.NOW,
+      domainName,
+    )
+
+    const delegationDTOs = delegations.map((d) => d.toDTO())
+
+    return delegationDTOs.map((d) => {
+      const person = this.getPersonByNationalId(fromNameInfo, d.fromNationalId)
+
+      return {
+        ...d,
+        fromName: person?.name ?? d.fromName ?? UNKNOWN_NAME,
+      }
+    })
+  }
+
+  async findAllAvailableIncoming(user: User): Promise<MergedDelegationDTO[]> {
+    const feature = await this.featureFlagService.getValue(
+      Features.customDelegations,
+      false,
+      user,
+    )
+    if (!feature) {
+      return []
+    }
+
+    const { delegations, fromNameInfo } = await this.findAllIncoming(
+      user,
+      DelegationValidity.NOW,
+    )
+
+    const allowedScopes = await this.getClientAllowedScopes(user)
+
+    const allowedDelegations = delegations
+      .map((d) => {
+        d.delegationScopes = d.delegationScopes?.filter((s) =>
+          this.checkIfScopeAllowed(s, allowedScopes),
+        )
+        return d
+      })
+      .filter(
+        (d) =>
+          // The requesting client must have access to at least one scope for the delegation to be relevant.
+          d.delegationScopes && d.delegationScopes.length > 0,
+      )
+
+    const mergedDelegationDTOs = uniqBy(
+      allowedDelegations.map((d) => d.toMergedDTO()),
+      'fromNationalId',
+    )
```

Comment:
Should we also merge these delegations by `fromNationalId`?

Reply:
Not sure if we need that?

Reply:
We need this if/when we start supporting (third party?) clients which get consented access to scopes from more than one domain.